src/dag/write.ts

```diff
@@ -85,75 +85,81 @@ export class Write extends Read {
   }
 
   async commit(): Promise<void> {
-    await this.collectGarbage();
+    await this._collectGarbage();
     await this._tx.commit();
   }
 
-  async collectGarbage(): Promise<void> {
+  private async _collectGarbage(): Promise<void> {
     // We increment all the ref counts before we do all the decrements. This
     // is so that we do not remove an item that goes from 1 -> 0 -> 1
-    const newHeads: (Hash | undefined)[] = [];
-    const oldHeads: (Hash | undefined)[] = [];
+    const newHeads: Hash[] = [];
+    const oldHeads: Hash[] = [];
     for (const changedHead of this._changedHeads.values()) {
-      oldHeads.push(changedHead.old);
-      newHeads.push(changedHead.new);
+      changedHead.old && oldHeads.push(changedHead.old);
+      changedHead.new && newHeads.push(changedHead.new);
     }
 
+    const refCountCache: Map<Hash, number> = new Map();
+
     for (const n of newHeads) {
-      if (n !== undefined) {
-        await this.changeRefCount(n, 1);
-      }
+      await this._changeRefCount(n, 1, refCountCache);
     }
 
     for (const o of oldHeads) {
-      if (o !== undefined) {
-        await this.changeRefCount(o, -1);
-      }
+      await this._changeRefCount(o, -1, refCountCache);
     }
 
+    await this._applyRefCountCache(refCountCache);
+
     // Now we go through the mutated chunks to see if any of them are still orphaned.
     const ps = [];
     for (const hash of this._newChunks) {
-      const count = await this.getRefCount(hash);
+      const count = await this._getRefCount(hash, refCountCache);
       if (count === 0) {
-        ps.push(this.removeAllRelatedKeys(hash, false));
+        ps.push(this._removeAllRelatedKeys(hash));
       }
     }
     await Promise.all(ps);
   }
 
-  async changeRefCount(hash: Hash, delta: number): Promise<void> {
-    const oldCount = await this.getRefCount(hash);
-    const newCount = oldCount + delta;
+  private async _changeRefCount(
+    hash: Hash,
+    delta: number,
+    refCountCache: Map<Hash, number>,
+  ): Promise<void> {
+    const oldCount = await this._getRefCount(hash, refCountCache);
 
     if ((oldCount === 0 && delta === 1) || (oldCount === 1 && delta === -1)) {
       const meta = await this._tx.get(chunkMetaKey(hash));
       if (meta !== undefined) {
         assertMeta(meta);
-        const ps = meta.map(ref => this.changeRefCount(ref, delta));
+        const ps = meta.map(ref =>
+          this._changeRefCount(ref, delta, refCountCache),
+        );
         await Promise.all(ps);
       }
     }
 
-    if (newCount === 0) {
-      await this.removeAllRelatedKeys(hash, true);
-    } else {
-      await this.setRefCount(hash, newCount);
+    {
+      // Read again since the ref count might have changed.
```

Comment:
This line is pretty scary to me. we read a value, make a decision base don that value (whether to descend) but then read it again because it might have changed?? 😬😬😬

Can I propose an alternate? What if we treat this like a breadth-first problem. Starting at the root:

1. collect all the ref changes to unique hashes in memory (so like if two different heads both reference same hash, then at this phase we would know that that hash is +2).
2. In parallel, execute all those changes.
3. For any hashes that went 0->1 or 1->0 do 1 again.

Reply:
(alternately, I think the previous solution with a per-hash lock was also clearly correct, which I'm less sure about here).